### PR TITLE
[feat][connector] Support 4paradigm/OpenMLDB jdbc sink connector

### PR DESCRIPTION
**Pull request by vagetablechicken**
### Motivation

Support sink to OpenMLDB.

### Modifications

Add a module 'openmldb 'in pulsar-io/jdbc

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

If we want to test this connector, we need to start an OpenMLDB cluster. So it can't be tested simply. But jdbc have tests.

### Does this pull request potentially affect one of the following parts:

No

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [x] `doc` 
(Your PR contains doc changes)

### CODE REVIEW

**Review by RobertIndie — CHANGES_REQUESTED** [review]
Hi. Thanks for your contribution.

Could you add the integration test like other JDBC connectors? Like [here](https://github.com/apache/pulsar/blob/master/tests/integration/src/test/java/org/apache/pulsar/tests/integration/io/sinks/JdbcPostgresSinkTester.java).
Could we add a container warp based on the image `4pdosc/openmldb:0.4.4` ?

Please also add this connector here : https://github.com/apache/pulsar/blob/8a67a383fb5e53f947d0513fc12656dd37961690/distribution/io/src/assemble/io.xml#L52-L56

**Comment by vagetablechicken** [comment]
assemble/io.xml fixed. And I'll add integration test in the next pr.
Plz help me to pass the ci. Need rerun ci? @RobertIndie

File: pulsar-io/jdbc/openmldb/pom.xml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    under the License.
+
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <parent>
+    <artifactId>pulsar-io-jdbc</artifactId>
+    <groupId>org.apache.pulsar</groupId>
+    <version>2.10.0-SNAPSHOT</version>
```
Comment (RobertIndie) [inline pulsar-io/jdbc/openmldb/pom.xml:27]:
```suggestion
    <version>2.11.0-SNAPSHOT</version>
```

**Comment by RobertIndie** [comment]
While testing, I found this exception when starting the connector. Could you please take a look? @vagetablechicken 
```
2022-04-11T15:58:55,371+0800 [function-timer-thread-17-1] ERROR org.apache.pulsar.functions.runtime.RuntimeSpawner - public/default/jdbc-openmldb-sink Function Container is dead with following exception. Restarting.
java.lang.UnsatisfiedLinkError: 'long com._4paradigm.openmldb.sql_router_sdkJNI.new_SQLRouterOptions()'
        at com._4paradigm.openmldb.sql_router_sdkJNI.new_SQLRouterOptions(Native Method) ~[?:?]
        at com._4paradigm.openmldb.SQLRouterOptions.<init>(SQLRouterOptions.java:56) ~[?:?]
        at com._4paradigm.openmldb.sdk.impl.SqlClusterExecutor.<init>(SqlClusterExecutor.java:66) ~[?:?]
        at com._4paradigm.openmldb.sdk.impl.SqlClusterExecutor.<init>(SqlClusterExecutor.java:90) ~[?:?]
        at com._4paradigm.openmldb.jdbc.SQLDriver.connect(SQLDriver.java:54) ~[?:?]
        at java.sql.DriverManager.getConnection(DriverManager.java:677) ~[java.sql:?]
        at java.sql.DriverManager.getConnection(DriverManager.java:189) ~[java.sql:?]
        at org.apache.pulsar.io.jdbc.JdbcAbstractSink.open(JdbcAbstractSink.java:93) ~[?:?]
        at org.apache.pulsar.functions.instance.JavaInstanceRunnable.setupOutput(JavaInstanceRunnable.java:857) ~[classes/:?]
        at org.apache.pulsar.functions.instance.JavaInstanceRunnable.setup(JavaInstanceRunnable.java:223) ~[classes/:?]
        at org.apache.pulsar.functions.instance.JavaInstanceRunnable.run(JavaInstanceRunnable.java:258) ~[classes/:?]
        at java.lang.Thread.run(Thread.java:829) [?:?]

```

<details>
<summary>Full exception context</summary>
<code><pre>
2022-04-11T15:58:13,069+0800 [main] INFO  org.apache.pulsar.functions.utils.io.ConnectorUtils - Searching for connectors in /Users/aaronrobert/codebase/pulsar/connectors
2022-04-11T15:58:13,097+0800 [main] WARN  org.apache.pulsar.functions.utils.io.ConnectorUtils - Connectors archive directory not found
2022-04-11T15:58:13,128+0800 [main] INFO  org.apache.pulsar.common.nar.NarUnpacker - Created directory /var/folders/sx/ws7wtlmn3t1bhrdlsfbmlv6r0000gn/T/pulsar_localrunner_nars_3526555380500249181/pulsar-nar/pulsar-io-jdbc-openmldb-2.11.0-SNAPSHOT.nar-unpacked
2022-04-11T15:58:15,559+0800 [main] INFO  org.apache.pulsar.common.nar.NarUnpacker - Extracting /Users/aaronrobert/codebase/pulsar/pulsar-io/jdbc/openmldb/target/pulsar-io-jdbc-openmldb-2.11.0-SNAPSHOT.nar to /var/folders/sx/ws7wtlmn3t1bhrdlsfbmlv6r0000gn/T/pulsar_localrunner_nars_3526555380500249181/pulsar-nar/pulsar-io-jdbc-openmldb-2.11.0-SNAPSHOT.nar-unpacked/ABc4BPN6ZjOQMHeC_JQ1xg
2022-04-11T15:58:19,304+0800 [main] WARN  io.netty.resolver.dns.DnsServerAddressStreamProviders - Can not find io.netty.resolver.dns.macos.MacOSDnsServerAddressStreamProvider in the classpath, fallback to system defaults. This may result in incorrect DNS resolutions on MacOS.
2022-04-11T15:58:24,580+0800 [main] INFO  org.apache.pulsar.functions.runtime.RuntimeSpawner - public/default/jdbc-openmldb-sink-0 RuntimeSpawner starting function
2022-04-11T15:58:24,583+0800 [main] INFO  org.apache.pulsar.common.nar.FileUtils - Jar file /Users/aaronrobert/codebase/pulsar/pulsar-io/jdbc/openmldb/target/pulsar-io-jdbc-openmldb-2.11.0-SNAPSHOT.nar contains META-INF/bundled-dependencies, it may be a NAR file
2022-04-11T15:58:24,583+0800 [main] INFO  org.apache.pulsar.functions.runtime.thread.ThreadRuntime - Trying Loading file as NAR file: /Users/aaronrobert/codebase/pulsar/pulsar-io/jdbc/openmldb/target/pulsar-io-jdbc-openmldb-2.11.0-SNAPSHOT.nar
2022-04-11T15:58:25,331+0800 [main] INFO  org.apache.pulsar.functions.runtime.thread.ThreadRuntime - Initialize function class loader for function jdbc-openmldb-sink at function cache manager, functionClassLoader: org.apache.pulsar.common.nar.NarClassLoader[/var/folders/sx/ws7wtlmn3t1bhrdlsfbmlv6r0000gn/T/pulsar_localrunner_nars_3526555380500249181/pulsar-nar/pulsar-io-jdbc-openmldb-2.11.0-SNAPSHOT.nar-unpacked/ABc4BPN6ZjOQMHeC_JQ1xg]
2022-04-11T15:58:25,361+0800 [main] INFO  org.apache.pulsar.functions.runtime.thread.ThreadRuntime - ThreadContainer starting function with instanceId 0 functionId 9c836a56-83a4-416f-b9b7-bbabedad787f namespace default
2022-04-11T15:58:55,371+0800 [function-timer-thread-17-1] ERROR org.apache.pulsar.functions.runtime.RuntimeSpawner - public/default/jdbc-openmldb-sink Function Container is dead with following exception. Restarting.
java.lang.UnsatisfiedLinkError: 'long com._4paradigm.openmldb.sql_router_sdkJNI.new_SQLRouterOptions()'
        at com._4paradigm.openmldb.sql_router_sdkJNI.new_SQLRouterOptions(Native Method) ~[?:?]
        at com._4paradigm.openmldb.SQLRouterOptions.<init>(SQLRouterOptions.java:56) ~[?:?]
        at com._4paradigm.openmldb.sdk.impl.SqlClusterExecutor.<init>(SqlClusterExecutor.java:66) ~[?:?]
        at com._4paradigm.openmldb.sdk.impl.SqlClusterExecutor.<init>(SqlClusterExecutor.java:90) ~[?:?]
        at com._4paradigm.openmldb.jdbc.SQLDriver.connect(SQLDriver.java:54) ~[?:?]
        at java.sql.DriverManager.getConnection(DriverManager.java:677) ~[java.sql:?]
        at java.sql.DriverManager.getConnection(DriverManager.java:189) ~[java.sql:?]
        at org.apache.pulsar.io.jdbc.JdbcAbstractSink.open(JdbcAbstractSink.java:93) ~[?:?]
        at org.apache.pulsar.functions.instance.JavaInstanceRunnable.setupOutput(JavaInstanceRunnable.java:857) ~[classes/:?]
        at org.apache.pulsar.functions.instance.JavaInstanceRunnable.setup(JavaInstanceRunnable.java:223) ~[classes/:?]
        at org.apache.pulsar.functions.instance.JavaInstanceRunnable.run(JavaInstanceRunnable.java:258) ~[classes/:?]
        at java.lang.Thread.run(Thread.java:829) [?:?]
2022-04-11T15:58:55,395+0800 [function-timer-thread-17-1] INFO  org.apache.pulsar.functions.runtime.thread.ThreadRuntime - Unloading JAR files for instanceId 0 functionId 9c836a56-83a4-416f-b9b7-bbabedad787f namespace default
2022-04-11T15:58:55,397+0800 [function-timer-thread-17-1] INFO  org.apache.pulsar.common.nar.FileUtils - Jar file /Users/aaronrobert/codebase/pulsar/pulsar-io/jdbc/openmldb/target/pulsar-io-jdbc-openmldb-2.11.0-SNAPSHOT.nar contains META-INF/bundled-dependencies, it may be a NAR file
2022-04-11T15:58:55,397+0800 [function-timer-thread-17-1] INFO  org.apache.pulsar.functions.runtime.thread.ThreadRuntime - Trying Loading file as NAR file: /Users/aaronrobert/codebase/pulsar/pulsar-io/jdbc/openmldb/target/pulsar-io-jdbc-openmldb-2.11.0-SNAPSHOT.nar
2022-04-11T15:58:55,920+0800 [function-timer-thread-17-1] INFO  org.apache.pulsar.functions.runtime.thread.ThreadRuntime - Initialize function class loader for function jdbc-openmldb-sink at function cache manager, functionClassLoader: org.apache.pulsar.common.nar.NarClassLoader[/var/folders/sx/ws7wtlmn3t1bhrdlsfbmlv6r0000gn/T/pulsar_localrunner_nars_3526555380500249181/pulsar-nar/pulsar-io-jdbc-openmldb-2.11.0-SNAPSHOT.nar-unpacked/ABc4BPN6ZjOQMHeC_JQ1xg]
^C2022-04-11T15:58:59,284+0800 [Thread-0] INFO  org.apache.pulsar.functions.runtime.thread.ThreadRuntime - Unloading JAR files for instanceId 0 functionId 9c836a56-83a4-416f-b9b7-bbabedad787f namespace default
2022-04-11T15:58:59,285+0800 [Thread-0] INFO  org.apache.pulsar.client.impl.PulsarClientImpl - Client closing. URL: pulsar://localhost:6650
➜  openmldb git:(openmldb-connector) ✗ 2022-04-11T15:59:00,930+0800 [function-timer-thread-17-1] INFO  org.apache.pulsar.functions.runtime.thread.ThreadRuntime - ThreadContainer starting function with instanceId 0 functionId 9c836a56-83a4-416f-b9b7-bbabedad787f namespace default

</pre></code>
</details>

**Review by RobertIndie — APPROVED** [review]

File: site2/docs/io-jdbc-sink.md [hunk 2]
```diff
@@ -15,16 +15,16 @@ The configuration of all JDBC sink connectors has the following properties.
...
-| `batchSize` | int|false | 200 | The batch size of updates made to the database. |
+| Name        | Type   | Required | Default            | Description                                                                                                          |
+|-------------|--------|----------|--------------------|----------------------------------------------------------------------------------------------------------------------|
+| `userName`  | String | false    | " " (empty string) | The username used to connect to the database specified by `jdbcUrl`.<br><br>**Note: `userName` is case-sensitive.**  |
+| `password`  | String | false    | " " (empty string) | The password used to connect to the database specified by `jdbcUrl`. <br><br>**Note: `password` is case-sensitive.** |
+| `jdbcUrl`   | String | true     | " " (empty string) | The JDBC URL of the database to which the connector connects.                                                        |
+| `tableName` | String | true     | " " (empty string) | The name of the table to which the connector writes.                                                                 |
```
Comment (momo-jun) [inline site2/docs/io-jdbc-sink.md:23]:
```suggestion
| `tableName` | String | true     | " " (empty string) | The name of the table that the connector writes to.                                                                 |
```

File: site2/docs/io-jdbc-sink.md [hunk 3]
```diff
@@ -15,16 +15,16 @@ The configuration of all JDBC sink connectors has the following properties.
...
+| `nonKey`    | String | false    | " " (empty string) | A comma-separated list contains the fields used in updating events.                                                  |
+| `key`       | String | false    | " " (empty string) | A comma-separated list contains the fields used in `where` condition of updating and deleting events.                |
```
Comment (Anonymitaet) [inline site2/docs/io-jdbc-sink.md:25]: same

**Comment by vagetablechicken** [comment]
> While testing, I found this exception when starting the connector. Could you please take a look? @vagetablechicken
> 
> ```
> 2022-04-11T15:58:55,371+0800 [function-timer-thread-17-1] ERROR org.apache.pulsar.functions.runtime.RuntimeSpawner - public/default/jdbc-openmldb-sink Function Container is dead with following exception. Restarting.
> java.lang.UnsatisfiedLinkError: 'long com._4paradigm.openmldb.sql_router_sdkJNI.new_SQLRouterOptions()'
>         at com._4paradigm.openmldb.sql_router_sdkJNI.new_SQLRouterOptions(Native Method) ~[?:?]
>         at com._4paradigm.openmldb.SQLRouterOptions.<init>(SQLRouterOptions.java:56) ~[?:?]
>         at com._4paradigm.openmldb.sdk.impl.SqlClusterExecutor.<init>(SqlClusterExecutor.java:66) ~[?:?]
>         at com._4paradigm.openmldb.sdk.impl.SqlClusterExecutor.<init>(SqlClusterExecutor.java:90) ~[?:?]
>         at com._4paradigm.openmldb.jdbc.SQLDriver.connect(SQLDriver.java:54) ~[?:?]
>         at java.sql.DriverManager.getConnection(DriverManager.java:677) ~[java.sql:?]
>         at java.sql.DriverManager.getConnection(DriverManager.java:189) ~[java.sql:?]
>         at org.apache.pulsar.io.jdbc.JdbcAbstractSink.open(JdbcAbstractSink.java:93) ~[?:?]
>         at org.apache.pulsar.functions.instance.JavaInstanceRunnable.setupOutput(JavaInstanceRunnable.java:857) ~[classes/:?]
>         at org.apache.pulsar.functions.instance.JavaInstanceRunnable.setup(JavaInstanceRunnable.java:223) ~[classes/:?]
>         at org.apache.pulsar.functions.instance.JavaInstanceRunnable.run(JavaInstanceRunnable.java:258) ~[classes/:?]
>         at java.lang.Thread.run(Thread.java:829) [?:?]
> ```
> 
> Full exception context

Fixed, use openmldb-native allinone version.

**Review by Anonymitaet — COMMENTED** [review]
LGTM from the tech writing pespective

**Review by Anonymitaet — APPROVED** [review]

**Review by momo-jun — APPROVED** [review]
LGTM.

**Review by codelipenghui — APPROVED** [review]

**Comment by codelipenghui** [comment]
@vagetablechicken could you please help check the failed `OWASP Dependency Check` test? I'm not sure if it was introduced by new dependencies, maybe you can try to rebase to the master branch first.

**Comment by RobertIndie** [comment]
The `openmldb-jdbc` depends on log4j 1.x and it causes the OWASP dependency check to fail. We need a new release of openmldb-jdbc to fix this security issue. This PR disables OWASP check for the opemmldb connector for now. And it will be reenabled until this security issue has been fixed.

Besides, the [current owasp dependency check failure](https://github.com/apache/pulsar/runs/6018518487?check_suite_focus=true) is not related to this PR.

**Pull request merged**